linker.ts

```diff
@@ -1,72 +1,132 @@
 import assert from 'assert';
 import { keccak256 } from 'js-sha3';
+import { isNil, isObject } from './common/helpers';
 
-function libraryHashPlaceholder (input) {
-  return '$' + keccak256(input).slice(0, 34) + '$';
+/**
+ * Newer versions of the compiler use hashed names instead of the string values.
+ *
+ * This helper makes the conversion between name and hash at the target length
+ * for replacement.
+ *
+ * @param libraryName
+ */
+function libraryHashPlaceholder (libraryName) {
+  return `$${keccak256(libraryName).slice(0, 34)}$`;
 }
 
-const linkBytecode = function (bytecode, libraries) {
+/**
+ * Used to support legacy library names and the current hash library names,
+ * replaces the names with the provided hex address.
+ *
+ * @param bytecode
+ * @param hexAddress
+ * @param labelName
+ */
+function replacePlaceholderLabelWithHexAddress (bytecode, hexAddress, labelName) {
+  // truncate to 36 characters, why, I assume this is because legacy names
+  // only had a length of 36 characters?
+  const truncatedName = labelName.slice(0, 36);
+
+  const prefix = '__';
+  const postFix = `${Array(37 - truncatedName.length).join('_')}__`;
+  const libLabel = `${prefix}${truncatedName}${postFix}`;
+
+  while (bytecode.indexOf(libLabel) >= 0) {
+    bytecode = bytecode.replace(libLabel, hexAddress);
+  }
+
+  return bytecode;
+}
+
+/**
+ * When using libraries, the resulting bytecode will contain placeholders for
+ * the real addresses of the referenced libraries. These have to be updated,
+ * via a process called linking, before deploying the contract.
+ *
+ * This method provides a simple helper for linking.
+ *
+ * @param {string} bytecode
+ * @param {string} libraries
+ * @returns {string} bytecode
+ */
```

Comment:
```suggestion
/**
 * Finds and all library placeholders in the provided bytecode and replaces them with actual addresses.
 * Supports both old- and new-style placeholders (even both in the same file).
 * See [Library Linking](https://docs.soliditylang.org/en/latest/using-the-compiler.html#library-linking)
 * for a full explanation of the linking process.
 * 
 * Example of a legacy placeholder: `__lib.sol:L_____________________________`
 * Example of a new-style placeholder: `__$cb901161e812ceb78cfe30ca65050c4337$__`
 *
 * @param {string} bytecode Hex-encoded bytecode string. All 40-byte substrings starting and ending with
 *     `__` will be interpreted as placeholders.
 * @param {Map<string, string>} libraries Mapping between fully qualified library names and the hex-encoded
 *     addresses they should be replaced with. Addresses shorter than 40 characters are automatically padded with zeros.
 *
 * @returns {string} bytecode Hex-encoded bytecode string with placeholders replaced with addresses.
 *    Note that some placeholders may remain in the bytecode if `libraries` does not provide addresses for all of them.
 */
```

Reply:
Well, I used `Map<string, string>` mostly because I needed to put something there. I only now reviewed your other PR and see what types you actually used.